### ISSUE TITLE
Accompanying issue for notebook 7 (Uncertainty, regularization and the deep learning toolset)

### ISSUE THREAD

**Issue by crisostomi** [erodola/DLAI-s2-2022#21]
Comment this issue with your answers to exercise A and B of notebook 7 "Uncertainty, regularization, and the deep learning toolset".

At the beginning of your response put a big A or B to signal the exercise you are answering to. Like this:

# A
If you answer to both A and B please do it in the same comment and put at the beginning of each answer the big A or B.

**Comment by vitoc98** [comment]
# B
### Test 1:
I first tried with the fifteenth sample and the (flipped) tenth sample:
![image](https://user-images.githubusercontent.com/93674584/166301750-2de77a85-0c6c-4f30-bf77-49c0b606e1df.png)
Here are the predictions:
``0 truck
1 truck
2 truck
3 truck
4 truck
5 truck
6 truck
7 truck
8 car
9 car
10 car
11 car``

It would appear that there isn't much uncertainty in the classification of the first picture (the truck), while the system is a little conflicted about the second one, and even making it just a bit closer to the first picture confuses the system quite a bit
![image](https://user-images.githubusercontent.com/93674584/166303706-85f7f837-1051-4977-92fe-5a2ccf041755.png)

### Test 2
I also tried with two very different pictures, as in the fifth sample and again the (flipped) tenth sample:
![image](https://user-images.githubusercontent.com/93674584/166309875-601b8794-82e4-4114-9094-1f65fc48eb54.png)

Here predictions were all over the place, with the system initially mistaking the frog for a deer:
``
0 deer
1 deer
2 deer
3 deer
4 frog
5 frog
6 car
7 car
8 car
9 car
10 car
11 car
``

It's interesting how the system only figured out it was looking at a frog when it started turning into a car, makes me wonder a thing or two about the model, apparently more closely linking frogs to cars than deers, are frogs the transportation method of the future?

Jokes aside, here the neuron activation graph was a mess, the system is apparently really conflicted on whether the first picture depicts a frog or not, as it transitioned to the car things quickly quieted down:
![image](https://user-images.githubusercontent.com/93674584/166309745-d8b0ffef-2812-4a6f-9f48-d03ef1c10fc5.png)


### Test 3
I then tried with another frog, which was properly classified this time, allowing me to see how the system reacts to very different pictures being blended together (by the way the car was not flipped in this case):
![image](https://user-images.githubusercontent.com/93674584/166306254-b7bae50d-5f08-4729-a9f8-50f7bf04855a.png)
Nothing strange in the predictions this time:
``0 frog
1 frog
2 frog
3 frog
4 frog
5 frog
6 frog
7 frog
8 car
9 car
10 car
11 car``

It's interesting how the system was still labeling the eighth picture (among the ones blending between frog and car) as a frog, despite it clearly looking like a car, in fact already the seventh picture looked like a car to me, despite the system being pretty confident (not much overlap between the two classes) that that was a frog
![image](https://user-images.githubusercontent.com/93674584/166312369-8dc12d66-3481-4ccc-b720-4ded356c5d22.png)

P.S. By "eighth" i mean picture number 7, by "seventh" i mean picture number 6 and so on, as the numbers of the pictures in the system start from 0

**Comment by vitoc98** [comment]
# B
I worked on this with a colleague (basile.1760927@studenti.uniroma1.it)
### Test 1:
We first tried with the fifteenth sample and the (flipped) tenth sample:
![image](https://user-images.githubusercontent.com/93674584/166301750-2de77a85-0c6c-4f30-bf77-49c0b606e1df.png)
Here are the predictions:
``0 truck
1 truck
2 truck
3 truck
4 truck
5 truck
6 truck
7 truck
8 car
9 car
10 car
11 car``

It would appear that there isn't much uncertainty in the classification of the first picture (the truck), while the system is a little conflicted about the second one, and even making it just a bit closer to the first picture confuses the system quite a bit
![image](https://user-images.githubusercontent.com/93674584/166303706-85f7f837-1051-4977-92fe-5a2ccf041755.png)

### Test 2
We also tried with two very different pictures, as in the fifth sample and again the (flipped) tenth sample:
![image](https://user-images.githubusercontent.com/93674584/166309875-601b8794-82e4-4114-9094-1f65fc48eb54.png)

Here predictions were all over the place, with the system initially mistaking the frog for a deer:
``
0 deer
1 deer
2 deer
3 deer
4 frog
5 frog
6 car
7 car
8 car
9 car
10 car
11 car
``

It's interesting how the system only figured out it was looking at a frog when it started turning into a car, makes me wonder a thing or two about the model, apparently more closely linking frogs to cars than deers, are frogs the transportation method of the future?

Jokes aside, here the neuron activation graph was a mess, the system is apparently really conflicted on whether the first picture depicts a frog or not, as it transitioned to the car things quickly quieted down:
![image](https://user-images.githubusercontent.com/93674584/166309745-d8b0ffef-2812-4a6f-9f48-d03ef1c10fc5.png)


### Test 3
We then tried with another frog, which was properly classified this time, allowing us to see how the system reacts to very different pictures being blended together (by the way the car was not flipped in this case):
![image](https://user-images.githubusercontent.com/93674584/166306254-b7bae50d-5f08-4729-a9f8-50f7bf04855a.png)
Nothing strange in the predictions this time:
``0 frog
1 frog
2 frog
3 frog
4 frog
5 frog
6 frog
7 frog
8 car
9 car
10 car
11 car``

It's interesting how the system was still labeling the eighth picture (among the ones blending between frog and car) as a frog, despite it clearly looking like a car, in fact already the seventh picture looked like a car to me, despite the system being pretty confident (not much overlap between the two classes) that that was a frog
![image](https://user-images.githubusercontent.com/93674584/166312369-8dc12d66-3481-4ccc-b720-4ded356c5d22.png)

P.S. By "eighth" i mean picture number 7, by "seventh" i mean picture number 6 and so on, as the numbers of the pictures in the system start from 0

**Comment by alesspalma** [comment]
# B

### Test 1: from cat to dog
![newplot](https://user-images.githubusercontent.com/48835050/167033234-3a89474b-d28a-4706-862e-e78d41fcbf0b.png)

network predictions are: `0 cat 1 cat 2 cat 3 cat 4 cat 5 dog 6 dog 7 dog 8 dog 9 dog 10 dog 11 dog`, so nothing surprisely even if I would say that the poses of the two puppies are very similar, main differences here are the ears and the color of the fur (I suppose the shape of the ears can be a key local feature in recognizing dogs).

![newplot](https://user-images.githubusercontent.com/48835050/167034683-76a87838-204e-48e6-90cc-5fed9490e68c.png)

It turns out that the network is very uncertain when predicting cat, and less uncertain when the image starts to turn into a dog. Interestingly, classification on the 0-th image (pure cat) is more uncertain than the classification on images 1 to 4. A possible confusing element could be the color of the background, in fact when it starts to darken the predictions are more accurate.

### Test 2: from frog to plane
![newplot](https://user-images.githubusercontent.com/48835050/167036988-39efd705-3a5a-4d56-a532-643b059d5f03.png)

Here we have two very different images, network predictions are: `0 frog 1 frog 2 frog 3 frog 4 frog 5 frog 6 plane 7 plane 8 plane 9 plane 10 plane 11 plane`.

![newplot](https://user-images.githubusercontent.com/48835050/167037115-5bf2f5a9-6678-4677-8e03-9fa2443c6d00.png)

Uncertainty is very low, as expected, so I would say that wrt these two samples the network was almost perfect.

**Comment by sh3rlock14** [comment]
# A
*note*: I ran the experiments with `epoch_num=10` (I didn't take 1.5h to train all the model with the original epochs).

These are the results I obtained with default settings:

![Accuracy across 10 epochs](https://user-images.githubusercontent.com/28751714/168333114-8bce119c-8743-471a-9aa4-609a568a7e99.png)

* Can we mitigate the memorization effect augmenting the training dataset with 
some transformations?

It looks like so: I applied some DA (RandomGrayscale, RandomHorizontalFlip, RandomRotate) and the `BatchNorm` model not only outperformed the other models during training, but it was also able to generalize well on `test_data`.

![Accuracy across 10 epochs with DA](https://user-images.githubusercontent.com/28751714/168334712-dc9413c0-2c75-44c3-b4d9-efa846807939.png)

* What happen if we rise the dropout coefficient? And if we apply dropout also on the input layer? 

I set `p=0.5` and the results I obtained are not so distant *w.r.t* the default setting, although some models *e.g.* the `StdDropoutLeNet` reduced the gap between `train` and `test accuracy` by ~5 points. 

# B

As can be seen from the `preds`, the model cannot even get the right answers with the images untouched, confusing a 'dog' for a 'cat' and a 'bird' for a 'car'.

![Bird_Dog_Missclassifcation](https://user-images.githubusercontent.com/28751714/168336094-3c5f045a-b145-4b7b-af6a-d8832591321b.jpg)

For the 'original dog image' (and the early mixins), we can see that the network is quite confident in discriminating *""a dog* from *"a bird"*, while the reverse is not true (at least in this particular case). Based on the image appearances I could guess that this confusion for the *bird* is due to the lack of distinctive *bird features* such as the tail, the wings the streamlined body... such a representation could have been easily misjudged by a network that has "seen" only birds by a different P.O.V., *e.g.* from a certain distance, during flight, in a nest... 

![activations before softmax](https://user-images.githubusercontent.com/28751714/168336199-91bdfe1c-133e-4f1c-a135-a9b76f0e44bf.png)

![activations after softmax](https://user-images.githubusercontent.com/28751714/168336206-bb3b60bd-3c31-45af-8cec-81123b6096e9.png)
